rules/no-assign-parameter.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * @fileoverview Disallow assigning to function parameters.
+ * @author Simon Hajjar <simon.j.hajjar@gmail.com>
+ */
+
+'use strict';
+
+module.exports = {
+
+    meta: {
+
+        docs: {
+            recommended: true,
+            type: 'error',
+            description: 'Disallow assigning to function parameters'
+        },
+
+        schema: []
+    },
+
+    create: function (context) {
+
+        function inspectBody (body, node, params) {
+            for (let statement of body) {
+                if (statement['type'] === 'ExpressionStatement') {
+
+                    // Check for any assignment of the variables
+                    if (statement['expression']['type'] === 'AssignmentExpression' &&
+                        params.indexOf(statement['expression']['left']['name']) >= 0) {
+                        context.report ({
+                            node: node,
+                            message: 'Disallow assigning to function parameters'
+                        });
+                    }
+
+                    // Check for incrementing and decrementing
+                    if (statement['expression']['type'] === 'UpdateExpression' &&
+                        params.indexOf(statement['expression']['argument']['name']) >= 0) {
+                        context.report ({
+                            node: node,
+                            message: 'Disallow assigning to function parameters'
+                        });
+                    }
+                }
+
+                if (['ForStatement', 'WhileStatement', 'DoWhileStatement'].indexOf(statement['type']) >= 0) {
+                    inspectBody (statement['body']['body'], node, params);
```

Comment:
This works, but I'm not sure if it would be a better approach to catch each of these declarations were caught separately from FunctionDeclarations. Any suggestions on that?

Reply:
Your technique is fine. Although it might not catch all issues, it assures that there will be no false positives. Please make sure it
- Catches any top-level assignments
- Catches any assignments happening in any of the loops
- catches inside if & else statements (they could be blocks or single statements)
eg
```
if (true) {...}

if (true)
    boo = 100;
```
You need to account for both

Also use a simple recursion trick: take the inspection of If & Loop out into their own functions. Inside these functions, just add a condition that if another `if` exists, call inspectIf or if loop exists, call inspectLoop (recursive).
This ensures that no matter how many levels deep that assignment resides, it will still get caught because the rule traverses every level

```
function foo(uint x) {
  if (true) {
    do {
      x += 1;
    } while (x < 10);
  }
}
```

Add tests for all above^ and we're good to merge!

Hope I was able to clearly explain it? If not, please let me know